// typedoc-plugin-markdown-pages/dist/models/markdown-page-collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const markdown_page_1 = require("./markdown-page");
const path_utilities_1 = require("../utilities/path-utilities");
const path_1 = require("path");
const url_1 = require("url");
const filesystem_utilities_1 = require("../utilities/filesystem-utilities");
const page_utilities_1 = require("../utilities/page-utilities");
class MarkdownPageCollection extends markdown_page_1.MarkdownPage {
    constructor(logger, path, url) {
        if (!filesystem_utilities_1.isDirectory(path)) {
            throw new Error(`Markdown page collection path "${path}" is not a directory.`);
        }
        super(path, url);
        this._logger = logger;
        this.path = path;
        this.url = url;
        this.children = [];
        this.title = page_utilities_1.parsePageTitleFromPath(path);
    }
    readContents() {
        try {
            const childItems = filesystem_utilities_1.getDirectoryContents(this.path);
            for (const childItem of childItems) {
                let childPage;
                const childPath = path_1.join(this.path, childItem);
                const isChildADirectory = filesystem_utilities_1.isDirectory(childPath);
                const childUrl = this._getItemUrl(childItem, isChildADirectory);
                if (isChildADirectory) {
                    childPage = new MarkdownPageCollection(this._logger, childPath, childUrl);
                }
                else {
                    childPage = new markdown_page_1.MarkdownPage(childPath, childUrl);
                }
                childPage.readContents();
                if (["index.md", "readme.md"].includes(childItem.toLowerCase())) {
                    this.contents = childPage.contents;
                }
                else if (page_utilities_1.isIndexChildPage(childItem)) {
                    this.contents = "";
                    this.url = childPage.url;
                    this.children.unshift(childPage);
                }
                else {
                    this.children.push(childPage);
                }
            }
            if (!this.contents) {
                this._generateIndexContents();
            }
        }
        catch (e) {
            throw new Error(`Failed to read page collection contents. ${e}`);
        }
    }
    log() {
        this._recursiveLog(this, 0);
    }
    _getItemUrl(sourceItemName, isDirectory) {
        const urlPathToCollection = path_utilities_1.getDirectoryName(this.url)
            .replace(/\/?$/, "/"); // ensure last character is a slash
        if (isDirectory) {
            const newDir = url_1.resolve(urlPathToCollection, sourceItemName)
                .replace(/\/?$/, "/"); // ensure last character is a slash
            return url_1.resolve(newDir, "index.html");
        }
        else {
            const fileExtension = path_utilities_1.getFileExtension(sourceItemName);
            let sourceItemNameWithoutExtension = sourceItemName.slice(0, (1 + fileExtension.length) * -1);
            if (page_utilities_1.isIndexChildPage(sourceItemNameWithoutExtension)) {
                sourceItemNameWithoutExtension = sourceItemNameWithoutExtension.substr(1);
            }
            return url_1.resolve(urlPathToCollection, sourceItemNameWithoutExtension) + ".html";
        }
    }
    _recursiveLog(page, depth) {
        const isCollection = !!page.children;
        if (depth === 0) {
            this._logger.verbose(page.title);
        }
        else {
            let prefix = "";
            for (let i = 0; i < depth; i++) {
                prefix += " ";
            }
            prefix += "|__";
            this._logger.verbose(`${prefix}${page.title}`);
        }
        if (isCollection) {
            for (const child of page.children) {
                const newDepth = depth + Math.floor(page.title.length / 2) + (depth === 0 ? 0 : 3);
                this._recursiveLog(child, newDepth);
            }
        }
    }
    _generateIndexContents() {
        try {
            this._logger.verbose(`Generating index contents for page "${this.title}"...`);
            let contents = "## Contents";
            for (const child of this.children) {
                const childUrl = path_utilities_1.getItemNameFromPath(child.url);
                contents += `\n- [${child.title}](${childUrl})`;
            }
            this.contents = contents;
            this._logger.verbose(this.contents);
        }
        catch (e) {
            const errorMessage = `Failed to generate index contents. ${e}`;
            this._logger.error(errorMessage);
            throw new Error(errorMessage);
        }
    }
}
exports.MarkdownPageCollection = MarkdownPageCollection;
