// typedoc-plugin-markdown-pages/dist/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const declaration_1 = require("typedoc/dist/lib/utils/options/declaration");
const constants_1 = require("./constants");
exports.LABEL_OPTION = new declaration_1.OptionDeclaration({
    component: constants_1.PLUGIN_NAME,
    help: "Markdown Pages Plugin: The navigation label for markdown pages.",
    name: "mdPagesLabel",
    type: declaration_1.ParameterType.String,
});
exports.OUTPUT_DIR_NAME_OPTION = new declaration_1.OptionDeclaration({
    component: constants_1.PLUGIN_NAME,
    help: "Markdown Pages Plugin: The name of the directory pages will be output to",
    name: "mdPagesOutputDirName",
    type: declaration_1.ParameterType.String,
});
exports.SOURCE_PATH_OPTION = new declaration_1.OptionDeclaration({
    component: constants_1.PLUGIN_NAME,
    help: "Markdown Pages Plugin: The path to the directory where pages will be read from",
    name: "mdPagesSourcePath",
    type: declaration_1.ParameterType.String,
});
